druid/src/widget_id.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::atomic::{AtomicU32, Ordering};
+
+use crate::kurbo::Size;
+
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget, WindowId,
+};
+
+/// The target of a command.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Target {
+    Window(WindowId),
+    Widget(WidgetId),
+}
+
+impl Target {
+    pub(crate) fn is_window(self) -> bool {
+        match self {
+            Target::Window(_) => true,
+            _ => false,
+        }
+    }
+}
+
+impl From<WindowId> for Target {
+    fn from(id: WindowId) -> Target {
+        Target::Window(id)
+    }
+}
+
+impl From<WidgetId> for Target {
+    fn from(id: WidgetId) -> Target {
+        Target::Widget(id)
+    }
+}
+
+// We might decide to clean this up; it's to make it easy to offer different
+// arguments to submit_command.
+pub trait IntoOptTarget {
+    fn into_opt_target(self) -> Option<Target>;
+}
+
+impl IntoOptTarget for Target {
+    fn into_opt_target(self) -> Option<Target> {
+        Some(self)
+    }
+}
+
+impl IntoOptTarget for Option<Target> {
+    fn into_opt_target(self) -> Option<Target> {
+        self
+    }
+}
+
+impl IntoOptTarget for WindowId {
+    fn into_opt_target(self) -> Option<Target> {
+        Some(Target::Window(self))
+    }
+}
+
+impl IntoOptTarget for WidgetId {
+    fn into_opt_target(self) -> Option<Target> {
+        Some(Target::Widget(self))
+    }
+}
+
+/// A unique identifier for a single widget.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct WidgetId(u32);
+
+static WIDGET_ID_COUNTER: AtomicU32 = AtomicU32::new(1);
+
+impl WidgetId {
+    /// Allocate a new, unique widget id.
+    ///
+    /// Do note that if we create 4 billion widgets there may be a collision.
+    pub(crate) fn next() -> WidgetId {
+        let id = WIDGET_ID_COUNTER.fetch_add(1, Ordering::Relaxed);
+        WidgetId(id)
+    }
+}
+
+/// A wrapper that adds an identity to an otherwise anonymous widget.
+pub struct IdentityWrapper<W> {
```

Comment:
Refresh my memory: is the sole purpose of this widget so that we can get access to the `WidgetId` for other purposes? (Say, stashing it in some other widget that is a companion?)

Reading through this I'm noticing a potential gotcha: this will only work if the `IdentityWrapper` itself has a pod, so for instance if you do like,

```rust
let (id, widget) = IdentityWrapper::wrap(TextBox::new());
row.add_child(widget.padding(5.0));
```

You will get assigned a new id?

Reply:
The `Padding` widget gets a new id, yes. If you have a wrapper widget that *doesn't* allocate a new WidgetPod (like `EnvScope` and likely lens stuff), then it needs to forward the `get_id` request. That is maybe an argument for raising it to public visibility.

Reply:
hmm, this makes me think we might want some sort of general purpose `child(&self) -> Option<&dyn Widget>` method on `Widget`; and if we do that it would be really nice if we had `as_any` so we could downcast to concrete widgets. 😬

Reply:
I've experimented with that but have always run into problems of trying to be too fancy with the Rust type system, like having object safety problems. But if you can make it work, I agree it would be helpful!